### PR TITLE
add events for score set + reward settlement

### PR DESCRIPTION
**Pull request by kpeluso**

### CODE REVIEW

File: x/emissions/types/events.go [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+// 			}
+
+// 			updatedEvent.Attributes[j] = updatedAttr
+// 		}
+
+// 		updatedEvents[i] = updatedEvent
+// 	}
+
+// 	return updatedEvents
+// }
```
Comment (br4e) [inline x/emissions/types/events.go:462]: probably don't want all the commented code?

File: x/emissions/types/events.go [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeInfererRewardsSet, blockHeight, rewards))
+}
+
+func EmitNewForecasterRewardsSetEvent(ctx sdk.Context, blockHeight BlockHeight, rewards []TaskReward) {
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeForecasterRewardsSet, blockHeight, rewards))
+}
+
+func EmitNewReputererAndDelegaterRewardsSetEvent(ctx sdk.Context, blockHeight BlockHeight, rewards []TaskReward) {
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeReputerAndDelegatorRewardsSet, blockHeight, rewards))
+}
```
Comment (br4e) [inline x/emissions/types/events.go:50]: include topic?

File: x/emissions/types/events.go [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+	EventTypeInfererScoresSet              = "inferer_scores_set"
+	EventTypeForecasterScoresSet           = "forecaster_scores_set"
+	EventTypeReputerScoresSet              = "reputer_scores_set"
+	EventTypeInfererRewardsSet             = "inferer_rewards_set"
+	EventTypeForecasterRewardsSet          = "forecaster_rewards_set"
+	EventTypeReputerAndDelegatorRewardsSet = "reputer_and_delegator_rewards_set"
+
+	AttributeKeyTopicId     = "topic_id"
+	AttributeKeyBlockHeight = "block_height"
+	AttributeKeyAddres      = "address"
```
Comment (br4e) [inline x/emissions/types/events.go:21]: `AttributeKeyAddress`

File: x/emissions/types/events.go [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+	scoreValues := make([]alloraMath.Dec, len(scores))
+	for i, score := range scores {
+		addresses[i] = score.Address
+		scoreValues[i] = score.Score
+	}
+	return sdk.NewEvent(
+		eventType,
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(topicId)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(blockHeight)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatStringsForEventAttribute(addresses)),
```
Comment (br4e) [inline x/emissions/types/events.go:70]: these keys are incorrect

File: x/emissions/types/events.go [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+	rewardValues := make([]alloraMath.Dec, len(rewards))
+	for i, reward := range rewards {
+		addresses[i] = reward.Address
+		rewardValues[i] = reward.Reward
+	}
+	return sdk.NewEvent(
+		eventType,
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(topicId)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(blockHeight)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatStringsForEventAttribute(addresses)),
```
Comment (br4e) [inline x/emissions/types/events.go:90]: same

File: x/emissions/types/events.go [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+		sdk.NewAttribute(AttributeKeyScores, FormatDecsForEventAttribute(scoreValues)),
+	)
+}
+
+func NewRewardsSetEventBase(eventType string, blockHeight BlockHeight, rewards []TaskReward) sdk.Event {
+	if len(rewards) == 0 {
+		return sdk.Event{}
+	}
```
Comment (br4e) [inline x/emissions/types/events.go:78]: is this what you want? To still emit an event? Maybe return nil and have the outer function handle that and not emit?

File: x/emissions/module/rewards/scores.go [hunk 7]
```diff
@@ -111,19 +111,26 @@ func GenerateReputerScores(
 		newScores = append(newScores, newScore)
 	}
 
+	types.EmitNewReputerScoresSetEvent(ctx, newScores)
```
Comment (br4e) [inline x/emissions/module/rewards/scores.go:114]: the scores aren't necessarily set here though right? Do we want to emit somewhere in the calling function?

File: x/emissions/module/rewards/scores.go [hunk 8]
```diff
@@ -157,6 +164,8 @@ func GenerateInferenceScores(ctx sdk.Context, keeper keeper.Keeper, topicId uint
 		}
 		newScores = append(newScores, newScore)
 	}
+
+	types.EmitNewInfererScoresSetEvent(ctx, newScores)
```
Comment (br4e) [inline x/emissions/module/rewards/scores.go:168]: same

File: x/emissions/module/rewards/scores.go [hunk 9]
```diff
@@ -234,6 +243,7 @@ func GenerateForecastScores(
 		newScores = append(newScores, newScore)
 	}
 
+	types.EmitNewForecasterScoresSetEvent(ctx, newScores)
```
Comment (br4e) [inline x/emissions/module/rewards/scores.go:246]: same

File: x/emissions/module/rewards/rewards.go [hunk 10]
```diff
@@ -455,9 +462,18 @@ func payoutRewards(
...
+				infererRewards = append(infererRewards, reward)
+			} else if reward.Type == types.WorkerForecastRewardType {
+				forecasterRewards = append(forecasterRewards, reward)
+			}
 		}
 	}
 
+	types.EmitNewInfererRewardsSetEvent(ctx, blockHeight, infererRewards)
+	types.EmitNewForecasterRewardsSetEvent(ctx, blockHeight, forecasterRewards)
+	types.EmitNewReputererAndDelegaterRewardsSetEvent(ctx, blockHeight, reputerAndDelegatorRewards)
```
Comment (br4e) [inline x/emissions/module/rewards/rewards.go:476]: same. Maybe just don't call them "set" events?

**Review by br4e — COMMENTED** [review]
Generally makes sense, Definitely some changes needed. You also mentioned you aren't sure if this is emitting events currently?

File: x/emissions/types/events.go (hunk 2 above)
Comment (kpeluso) [reply x/emissions/types/events.go:50]: already in the reward struct

File: x/emissions/module/rewards/rewards.go [hunk 11]
```diff
@@ -212,19 +212,20 @@ func GenerateRewardsDistributionByTopicParticipant(
...
-		return []TaskRewards{}, alloraMath.Dec{}, errors.Wrapf(err, "failed to get network loss bundle at block %d", blockHeight)
+		return []types.TaskReward{}, alloraMath.Dec{}, errors.Wrapf(err, "failed to get network loss bundle at block %d", blockHeight)
 	}
 
 	// Calculate and Set the reputer scores
 	reputerScores, err := GenerateReputerScores(ctx, k, topicId, blockHeight, *bundles)
 	if err != nil {
 		return nil, alloraMath.Dec{}, err
 	}
+	// ctx.EventManager().EmitTypedEvent()
```
[the remark this answers is not in the document]
Comment (kpeluso) [reply x/emissions/module/rewards/rewards.go:228]: no

File: x/emissions/module/rewards/rewards.go (hunk 10 above)
Comment (kpeluso) [reply x/emissions/module/rewards/rewards.go:476]: renamed: settled

File: x/emissions/module/rewards/scores.go (hunk 7 above)
Comment (kpeluso) [reply x/emissions/module/rewards/scores.go:114]: they're set just before. The event should be emitted when it was set under certain conditions, which I'd say is responsibility for context above. ofc we don't use the set function elsewhere... but events seem more tied to reward imo

File: x/emissions/types/events.go (hunk 6 above)
Comment (kpeluso) [reply x/emissions/types/events.go:78]: no I don't...

File: x/emissions/module/rewards/rewards.go (hunk 10 above)
Comment (br4e) [reply x/emissions/module/rewards/rewards.go]: There's a spelling error in reputer

**Review by br4e — APPROVED** [review]
One minor spelling note for reputer but looks good otherwise

**Pull request merged**